File: reverseList/reverseList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *p, *q, *pr;
        if(!head)  return head;
        p = head;
        q = NULL;
        while(p)
        {
            pr = p->next;
            p->next = q;
            q = p;
            p = pr;
        }
        return q;
    }
};

File: reverseList/reverseList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return head
        p = head
        q = None
        while p:
            pr = p.next
            p.next = q
            q = p
            p = pr
        return q
